diff --git a/include/private/android_filesystem_config.h b/include/private/android_filesystem_config.h
index 6521cbe..0f5bc40 100644
--- a/include/private/android_filesystem_config.h
+++ b/include/private/android_filesystem_config.h
@@ -63,6 +63,8 @@
 #define AID_NFC           1027  /* nfc subsystem */
 #define AID_SDCARD_R      1028  /* external storage read access */
 
+#define AID_AUDIT         1029  /* access to kernel audit stream */
+
 #define AID_SHELL         2000  /* adb and debug shell user */
 #define AID_CACHE         2001  /* cache access */
 #define AID_DIAG          2002  /* access to diagnostic resources */
@@ -121,6 +123,7 @@ static const struct android_id_info android_ids[] = {
     { "net_bt",    AID_NET_BT, },
     { "sdcard_r",  AID_SDCARD_R, },
     { "sdcard_rw", AID_SDCARD_RW, },
+    { "audit",    AID_AUDIT, },
     { "media_rw",  AID_MEDIA_RW, },
     { "vpn",       AID_VPN, },
     { "keystore",  AID_KEYSTORE, },
@@ -181,6 +184,7 @@ static struct fs_path_config android_dirs[] = {
 ** and will allow partial matches.
 */
 static struct fs_path_config android_files[] = {
+    { 00660, AID_ROOT,      AID_AUDIT,     "system/etc/audit/auditd.rules" },
     { 00440, AID_ROOT,      AID_SHELL,     "system/etc/init.goldfish.rc" },
     { 00550, AID_ROOT,      AID_SHELL,     "system/etc/init.goldfish.sh" },
     { 00440, AID_ROOT,      AID_SHELL,     "system/etc/init.trout.rc" },
@@ -213,6 +217,7 @@ static struct fs_path_config android_files[] = {
     { 06755, AID_ROOT,      AID_ROOT,      "system/xbin/procmem" },
     { 06755, AID_ROOT,      AID_ROOT,      "system/xbin/tcpdump" },
     { 04770, AID_ROOT,      AID_RADIO,     "system/bin/pppd-ril" },
+    { 04750, AID_ROOT,      AID_AUDIT,     "system/bin/auditctl" },
 		/* the following file is INTENTIONALLY set-uid, and IS included
 		 * in user builds. */
     { 06750, AID_ROOT,      AID_SHELL,     "system/bin/run-as" },
@@ -227,6 +232,10 @@ static struct fs_path_config android_files[] = {
     { 00750, AID_ROOT,      AID_SHELL,     "sbin/fs_mgr" },
     { 00640, AID_ROOT,      AID_SHELL,     "fstab.*" },
     { 00644, AID_ROOT,      AID_ROOT,       0 },
+
+    /* The following is used to make sure the audit socket has appropriate
+     * UID/GID permissions. */
+    { 00660, AID_AUDIT,     AID_AUDIT,    "/dev/audit" },
 };
 
 static inline void fs_config(const char *path, int dir,
diff --git a/rootdir/init.rc b/rootdir/init.rc
index bb7e4ea..5f6d83a 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -363,6 +363,18 @@ service adbd /sbin/adbd
 on property:ro.kernel.qemu=1
     start adbd
 
+# init does not like either the fork() or pipe() call made by auditd during daemonization with forks. The -n stops these calls from being made and failing.
+service auditd /system/bin/auditd -n
+    class core
+    socket audit stream 0660 audit audit
+    critical
+
+# Do a one-off run of auditctl to start with a default set of rules
+service auditctl /system/bin/auditctl -R /etc/audit/auditd.rules
+    class core
+    oneshot
+
+
 service servicemanager /system/bin/servicemanager
     class core
     user system
@@ -373,6 +385,7 @@ service servicemanager /system/bin/servicemanager
     onrestart restart surfaceflinger
     onrestart restart drm
 
+
 service vold /system/bin/vold
     class core
     socket vold stream 0660 root mount
